File: backend/src/main/java/com/sistemadesaude/backend/paciente/service/PacienteService.java
package com.sistemadesaude.backend.paciente.service;

import com.sistemadesaude.backend.paciente.dto.PacienteDTO;
import com.sistemadesaude.backend.paciente.dto.PacienteListDTO;

import java.util.List;

/**
 * Interface de serviço para operações relacionadas ao Paciente.
 */
public interface PacienteService {

    /**
     * Cria um novo paciente
     */
    PacienteDTO criarPaciente(PacienteDTO dto);

    /**
     * Busca paciente por ID
     */
    PacienteDTO buscarPacientePorId(Long id);

    /**
     * Atualiza um paciente
     */
    PacienteDTO atualizarPaciente(Long id, PacienteDTO dto);

    /**
     * Exclui um paciente
     */
    void excluirPaciente(Long id);

    /**
     * Busca pacientes por nome
     */
    List<PacienteListDTO> buscarPacientesPorNome(String nome);

    /**
     * Lista todos os pacientes
     */
    List<PacienteListDTO> listarTodosPacientes();

    /**
     * Verifica se o paciente está em vulnerabilidade
     */
    boolean verificarVulnerabilidade(Long id);

    // Métodos auxiliares
    List<PacienteDTO> listarTodos();
    PacienteDTO criar(PacienteDTO dto);
    PacienteDTO buscarPorId(Long id);
    List<PacienteDTO> buscarPorMultiplosCriterios(String termo);
    PacienteDTO buscarPorCpf(String cpf);
    PacienteDTO buscarPorCns(String cns);
}

File: frontend/src/services/paciente/PacienteService.ts
import axios from 'axios';
import { Paciente, PacienteList } from '../../types/paciente/Paciente';

const API_URL = '/api/pacientes';

/**
 * Serviço para operações relacionadas a pacientes.
 * Organizado de acordo com o domínio de Paciente no backend.
 */
export const PacienteService = {
  /**
   * Busca todos os pacientes.
   * @returns Lista de pacientes
   */
  listarTodos: async (): Promise<PacienteList[]> => {
    const response = await axios.get<PacienteList[]>(API_URL);
    return response.data;
  },

  /**
   * Busca pacientes pelo nome.
   * @param nome Nome ou parte do nome
   * @returns Lista de pacientes
   */
  buscarPorNome: async (nome: string): Promise<PacienteList[]> => {
    const response = await axios.get<PacienteList[]>(`${API_URL}/buscar?nome=${encodeURIComponent(nome)}`);
    return response.data;
  },

  /**
   * Busca um paciente pelo ID.
   * @param id ID do paciente
   * @returns Dados do paciente
   */
  buscarPorId: async (id: number): Promise<Paciente> => {
    const response = await axios.get<Paciente>(`${API_URL}/${id}`);
    return response.data;
  },

  /**
   * Cria um novo paciente.
   * @param paciente Dados do paciente
   * @returns Paciente criado
   */
  criar: async (paciente: Paciente): Promise<Paciente> => {
    const response = await axios.post<Paciente>(API_URL, paciente);
    return response.data;
  },

  /**
   * Atualiza os dados de um paciente.
   * @param id ID do paciente
   * @param paciente Novos dados do paciente
   * @returns Paciente atualizado
   */
  atualizar: async (id: number, paciente: Paciente): Promise<Paciente> => {
    const response = await axios.put<Paciente>(`${API_URL}/${id}`, paciente);
    return response.data;
  },

  /**
   * Exclui um paciente.
   * @param id ID do paciente
   */
  excluir: async (id: number): Promise<void> => {
    await axios.delete(`${API_URL}/${id}`);
  },

  /**
   * Verifica se o paciente está em situação de vulnerabilidade.
   * @param id ID do paciente
   * @returns true se o paciente estiver em situação de vulnerabilidade, false caso contrário
   */
  verificarVulnerabilidade: async (id: number): Promise<boolean> => {
    const response = await axios.get<boolean>(`${API_URL}/${id}/vulnerabilidade`);
    return response.data;
  }
};

export default PacienteService;